Core/include/Acts/TrackFitting/detail/GsfSmoothing.hpp

```diff
@@ -0,0 +1,274 @@
+/// [weight, parameters, covariance]. Therefore, it contains a MultiTrajectory
+/// and for now a std::map for the weights
+template <StatesType type>
+struct MultiTrajectoryProjector {
+  const MultiTrajectory &mt;
+  const std::map<std::size_t, double> &weights;
+
+  auto operator()(std::size_t idx) const {
+    const auto proxy = mt.getTrackState(idx);
+    switch (type) {
+      case StatesType::ePredicted:
+        return std::make_tuple(weights.at(idx), proxy.predicted(),
+                               std::optional{proxy.predictedCovariance()});
+      case StatesType::eFiltered:
+        return std::make_tuple(weights.at(idx), proxy.filtered(),
+                               std::optional{proxy.filteredCovariance()});
+      case StatesType::eSmoothed:
+        return std::make_tuple(weights.at(idx), proxy.smoothed(),
+                               std::optional{proxy.smoothedCovariance()});
+    }
+  }
+};
+
+/// @brief This function takes two MultiTrajectory objects and corresponding
+/// index lists (one of the backward pass, one of the forward pass), combines
+/// them, applies smoothing, and returns a new, single-component MultiTrajectory
+/// @tparam ReturnSmootedStates If set to true, the function returns not only
+/// combined MultiTrajectory, but also a std::vector contianing the
+/// component-wise smoothed states
+/// TODO this function does not handle outliers correctly at the moment I think
+/// TODO change std::vector< size_t > to boost::small_vector for better
+/// performance
+template <bool ReturnSmootedStates = false>
+auto smoothAndCombineTrajectories(
+    const MultiTrajectory &fwd, const std::vector<std::size_t> &fwdStartTips,
+    const std::map<std::size_t, double> &fwdWeights, const MultiTrajectory &bwd,
+    const std::vector<std::size_t> &bwdStartTips,
+    const std::map<std::size_t, double> &bwdWeights,
+    LoggerWrapper logger = getDummyLogger()) {
+  // This vector gets only filled if ReturnSmootedStates is true
+  std::vector<std::pair<const Surface *,
+                        std::vector<std::tuple<double, BoundVector,
+                                               std::optional<BoundSymMatrix>>>>>
+      smoothedStates;
+
+  // Use backward trajectory as basic trajectory, so that final trajectory is
+  // ordered correctly. We ensure also that they are unique.
+  std::vector<std::size_t> bwdTips = bwdStartTips;
+
+  // Ensures that the bwd tips are unique and do not contain MAX_SIZE which
+  // represents an invalid trajectory state
+  auto sort_unique_validate_bwd_tips = [&]() {
+    std::sort(bwdTips.begin(), bwdTips.end());
+    bwdTips.erase(std::unique(bwdTips.begin(), bwdTips.end()), bwdTips.end());
+
+    auto invalid_it = std::find(bwdTips.begin(), bwdTips.end(),
+                                std::numeric_limits<uint16_t>::max());
+    if (invalid_it != bwdTips.end()) {
+      bwdTips.erase(invalid_it);
+    }
+  };
+
+  sort_unique_validate_bwd_tips();
```

Comment:
That's not entirely according to our coding standards, but I don't mind too much.

Reply:
Should be a simple fix though. Up to you @benjaminhuth

Reply:
You mean the not-CamelCase or the use of the lambda?

Reply:
`snake_case`, yes. The lambda is fine!